### PR TITLE
authz: slow down primary key grow pace of "user_pending_permissions" table

### PR DESCRIPTION
**Pull request by unknwon**
### Context

We're using 32-bit integer (`int32` in Go / `int4` in Postgres) to store a user ID, and we are using [roaring bitmap](https://roaringbitmap.org/) as the efficient in-memory representation for a collection of user IDs, which only supports `uint32`. Therefore it was a natural choice to use `int4` as the column type for `user_pending_permissions` table (same as `users` table).

### What is the problem with existing approach?

Just like we use a primary key with auto-increment property for `users` table, we have the same use case for `user_pending_permissions`. The problem comes from that [Postgres still monotonically increases the primary key with failed inserts](https://dba.stackexchange.com/a/83351) (which based on our query, it becomes an upsert).

In our case, if there are lots of users with pending permissions, we're making lots of failed inserts with every call of `SetRepoPendingPermissions`. The grow pace of the primary key can be simply calculated as follows:

If a repository has 10k users (that are not yet having an account in the Sourcegraph instance, but have access to on the code host), 2,147,483,647 (the largest number a signed 32-bit integer can hold) would only support 214,748 calls of `SetRepoPendingPermissions`. It may seem a lot at a first glance, but imagine our background syncing process picks 10 repositories every minute, that gives us `214748 / 10 / 60 / 24 ~= 15` days, not to mention the initial syncing round would have a call to `SetRepoPendingPermissions` for every single repository.

### Why no customers have ran into this problem?

My hypothesis:

Most of users on the code host have onboarded to the Sourcegraph instance, which makes the primary key grow pace very slow, e.g. < 100 for every call of `SetRepoPendingPermissions`, that is about `21474800 / 10 / 60 / 24 ~= 1500` days to reach the largest number (background permissions was introduced about 18 months / 540 days ago, that leaves customers ~900 days).

### How the approach in this PR helps?

Instead of naively relying on upsert for massive existing rows (that causes fail inserts), we first load existing rows and only do inserts for newly discovered rows. This means if there are 20k users on the code host, the primary key should also only grow to around 20k.

### CODE REVIEW

**Comment by ryanslade** [comment]
Looking at the code now, thanks for the detailed summary of the changes!

File: enterprise/internal/database/perms_store.go [hunk 1]
```diff
@@ -565,25 +565,61 @@ func (s *PermsStore) SetRepoPendingPermissions(ctx context.Context, accounts *ex
...
-		// NOTE: Row-level locking is not needed here because we're creating stub rows and not modifying permissions.
-		q, err = insertUserPendingPermissionsBatchQuery(accounts, p)
+		// NOTE: The primary key of "user_pending_permissions" table is
+		//  auto-incremented, and it is monotonically growing even with upsert in
+		//  Postgres (i.e. the primary key is increased internally by one even if the row
+		//  exists). This means with large number of AccountIDs, the primary key will
+		//  grow very quickly every time we do an upsert, and soon reaching the largest
+		//  number an int4 (32-bit integer) can hold (2,147,483,647). Therefore, load
+		//  existing rows would help us only upsert rows that are newly discovered. See
+		//  NOTE in dozen lines below for why we do upsert not insert.
```
Comment (ryanslade) [inline enterprise/internal/database/perms_store.go:582]:
```suggestion
		//  NOTE in below for why we do upsert not insert.
```

The dozen will probably be outdated at some point :)

File: enterprise/internal/database/perms_store.go [hunk 2]
```diff
@@ -565,25 +565,61 @@ func (s *PermsStore) SetRepoPendingPermissions(ctx context.Context, accounts *ex
...
+		q = loadExistingUserPendingPermissionsBatchQuery(accounts, p)
+		bindIDsToIDs, err := txs.loadExistingUserPendingPermissionsBatch(ctx, q)
 		if err != nil {
-			return err
+			return errors.Wrap(err, "load existing user pending permissions")
```
Comment (ryanslade) [inline enterprise/internal/database/perms_store.go:586]:
```suggestion
			return errors.Wrap(err, "loading existing user pending permissions")
```

File: enterprise/internal/database/perms_store.go [hunk 3]
```diff
@@ -689,6 +735,33 @@ RETURNING id
...
+func loadExistingUserPendingPermissionsBatchQuery(accounts *extsvc.Accounts, p *authz.RepoPermissions) *sqlf.Query {
+	const format = `
+-- source: enterprise/internal/database/perms_store.go:loadExistingUserPendingPermissionsBatchQuery
+SELECT bind_id, id FROM user_pending_permissions
+WHERE
+	service_type = %s
+AND service_id = %s
+AND permission = %s
+AND object_type = %s
+AND bind_id IN (%s)
```
Comment (ryanslade) [inline enterprise/internal/database/perms_store.go:747]: I forget what the limit is, but if the number of ids is too large this will fail

**Review by ryanslade — APPROVED** [review]

File: enterprise/internal/database/perms_store.go (hunk 3 above)
Comment (bobheadxi) [reply enterprise/internal/database/perms_store.go:747]: A quick search turns up https://stackoverflow.com/questions/1009706/postgresql-max-number-of-parameters-in-in-clause which indicates in PostgresSQL there might not be a limit

Comment (ryanslade) [reply enterprise/internal/database/perms_store.go:747]: I know we've done some weird tricks in the past where we serialise everything to json object to get around a limit... Maybe that't no longer an issue

Comment (unknwon) [reply enterprise/internal/database/perms_store.go:747]: IIRC it's 65535, pretty sure other parts would start falling apart before here. We would want to some systematic audit for customers with super large user base.

**Review by bobheadxi — APPROVED** [review]

**Pull request merged**